cmd/kubeadm/app/util/apiclient/idempotency_test.go

```diff
@@ -237,3 +240,227 @@ func TestMutateConfigMapWithConflict(t *testing.T) {
+	type args struct {
+		client      clientset.Interface
+		clusterRole *rbac.ClusterRole
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "ClusterRole does not exist",
+			args: args{
+				client: fake.NewSimpleClientset(),
+				clusterRole: &rbac.ClusterRole{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "ClusterRole",
+						APIVersion: "v1",
+					},
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "foo",
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "ClusterRole exists",
+			args: args{
+				client: newMockClientForTest(t, "", "foo", "ClusterRole"),
+				clusterRole: &rbac.ClusterRole{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "ClusterRole",
+						APIVersion: "v1",
+					},
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "foo",
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "ClusterRole is invalid",
+			args: args{
+				client:      fake.NewSimpleClientset(),
+				clusterRole: nil,
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := CreateOrUpdateClusterRole(tt.args.client, tt.args.clusterRole); (err != nil) != tt.wantErr {
+				t.Errorf("CreateOrUpdateClusterRole() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestCreateOrUpdateClusterRoleBinding(t *testing.T) {
+	type args struct {
+		client             clientset.Interface
+		clusterRoleBinding *rbac.ClusterRoleBinding
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "ClusterRoleBinding does not exist",
+			args: args{
+				client: fake.NewSimpleClientset(),
+				clusterRoleBinding: &rbac.ClusterRoleBinding{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "ClusterRoleBinding",
+						APIVersion: "v1",
+					},
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "foo",
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "ClusterRoleBinding exists",
+			args: args{
+				client: newMockClientForTest(t, "", "foo", "ClusterRoleBinding"),
+				clusterRoleBinding: &rbac.ClusterRoleBinding{
+					TypeMeta: metav1.TypeMeta{
+						Kind:       "ClusterRoleBinding",
+						APIVersion: "v1",
+					},
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "foo",
+					},
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "ClusterRoleBinding is invalid",
+			args: args{
+				client:             newMockClientForTest(t, "", "foo", "ClusterRoleBinding"),
+				clusterRoleBinding: nil,
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := CreateOrUpdateClusterRoleBinding(tt.args.client, tt.args.clusterRoleBinding); (err != nil) != tt.wantErr {
+				t.Errorf("CreateOrUpdateClusterRoleBinding() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func newMockClientForTest(t *testing.T, namepsace string, name string, kind string) *fake.Clientset {
```

Comment:
BTW instead of using such a function it's simpler to use reactors.

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go#L838-L840
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/kubeconfig/kubeconfig_test.go#L936-L937

but the way you are doing it is also OK.